[rustc_ty_utils] Add the LLVM `noalias` parameter attribute to `drop_in_place` in certain cases.

pcwalton opened the pull request:
LLVM can make use of the `noalias` parameter attribute on the parameter to `drop_in_place` in areas like argument promotion. Because the Rust compiler fully controls the code for `drop_in_place`, it can soundly deduce parameter attributes on it. In the case of a value that has a programmer-defined Drop implementation, we know that the first thing `drop_in_place` will do is pass a pointer to the object to `Drop::drop`. `Drop::drop` takes `&mut`, so it must be guaranteed that there are no pointers to the object upon entering that function. Therefore, it should be safe to mark `noalias` there.

With this patch, we mark `noalias` only when the type is a value with a programmer-defined Drop implementation. This is possibly overly conservative, but I thought that proceeding cautiously was best in this instance.

pcwalton committed:
[rustc_ty_utils] Add the LLVM `noalias` parameter attribute to `drop_in_place` in certain cases.

LLVM can make use of the `noalias` parameter attribute on the parameter to
`drop_in_place` in areas like argument promotion. Because the Rust compiler
fully controls the code for `drop_in_place`, it can soundly deduce parameter
attributes on it. In the case of a value that has a programmer-defined Drop
implementation, we know that the first thing `drop_in_place` will do is pass a
pointer to the object to `Drop::drop`. `Drop::drop` takes `&mut`, so it must be
guaranteed that there are no pointers to the object upon entering that
function. Therefore, it should be safe to mark `noalias` there.

With this patch, we mark `noalias` only when the type is a value with a
programmer-defined Drop implementation. This is possibly overly conservative,
but I thought that proceeding cautiously was best in this instance.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-13`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3334342463/jobs/5517199189) [(plain)](https://github.com/rust-lang/rust/commit/4da9d39f608274ffe78c819a3ace7b6e7a6d8317/checks/9132412041/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

running 377 tests
i.....i..............i....i..ii.................i..ii......ii.i........i................ 88/377
..ii.................i............i..i.................i...iii........i..i......i.iiii.. 176/377
......i.iii....i..................i..Fi...i.....ii..i.ii..................ii............ 264/377
Some tests failed in compiletest suite=codegen mode=codegen host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
iiii.i...................
failures:


---- [codegen] src/test/codegen/noalias-box-off.rs stdout ----

error: verification with 'FileCheck' failed
status: exit status: 1
command: "/usr/lib/llvm-13/bin/FileCheck" "--input-file" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/noalias-box-off/noalias-box-off.ll" "/checkout/src/test/codegen/noalias-box-off.rs" "--allow-unused-prefixes" "--check-prefixes" "CHECK,NONMSVC"
stdout: none
--- stderr -------------------------------
/checkout/src/test/codegen/noalias-box-off.rs:6:15: error: CHECK-NOT: excluded string found in input
// CHECK-NOT: noalias
              ^
/checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/noalias-box-off/noalias-box-off.ll:12:68: note: found here
 tail call void @__rust_dealloc(i8* nonnull %0, i64 1, i64 1) #2, !noalias !2


Input file: /checkout/obj/build/x86_64-unknown-linux-gnu/test/codegen/noalias-box-off/noalias-box-off.ll


-dump-input=help explains the following input dump.
Input was:
<<<<<<
       .
       .
       .
       .
       7: %"unwind::libunwind::_Unwind_Context" = type { [0 x i8] } 
       8:  
       9: ; Function Attrs: nounwind nonlazybind uwtable 
      10: define void @box_should_not_have_noalias_if_disabled(i8* noundef nonnull align 1 %0) unnamed_addr #0 personality i32 (i32, i32, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*)* @rust_eh_personality { 
      11: start: 
      12:  tail call void @__rust_dealloc(i8* nonnull %0, i64 1, i64 1) #2, !noalias !2 
not:6                                                                        !~~~~~~     error: no match expected
      13:  ret void 
      14: } 
      15:  
      16: ; Function Attrs: nonlazybind uwtable 
      17: declare noundef i32 @rust_eh_personality(i32, i32 noundef, i64, %"unwind::libunwind::_Unwind_Exception"*, %"unwind::libunwind::_Unwind_Context"*) unnamed_addr #1 
       .
       .
>>>>>>
------------------------------------------
```

</details>

pcwalton committed:
Fix noalias box test

oli-obk commented on compiler/rustc_ty_utils/src/abi.rs:
```diff
@@ -276,6 +277,25 @@ fn adjust_for_rust_scalar<'tcx>(
         }
     }
+
+    // If this is the argument to `drop_in_place`, the contents of which we fully control as the
+    // compiler, then we may be able to mark that argument `noalias`. Currently, we're conservative
+    // and do so only if `drop_in_place` results in a direct call to the programmer's `drop` method.
+    // The `drop` method requires `&mut self`, so we're effectively just propagating the `noalias`
+    // guarantee from `drop` upward to `drop_in_place` in this case.
```
Why doesn't this fall out of the deduced parameter attribute analysis done on MIR? Can we make it happen there instead of special casing Abi computation further?

bjorn3 replied:
`deduced_param_attrs` only runs on user written functions, not on shims generated by the compiler. In addition it takes a `DefId`, but for `drop_in_place` the generated shim is unique for each generic argument.

pcwalton replied:
In addition to what @bjorn3 said, `deduced_param_attrs` is slower than doing the attribute assignment here, because it adds a query and has to write the results into the crate metadata. By doing the attribute assignment here, we can avoid that overhead for attributes that can be determined solely from the function identity and signature. This is the case for the pointer on `drop_in_place`—we know what the semantics of that function are immediately without having to inspect any MIR, and we can inspect the type of the parameter being dropped.

oli-obk commented:
Please also add comments on drop_in_place in libstd to make sure anyone touching it is aware they should also touch this logic

pcwalton commented:
Per discussions on Zulip, I'm going to change this to a draft while we wait on Miri results to figure out what the impact of this change will be.

pcwalton committed:
Apply `noalias`, `nonnull`, `dereferenceable`, and `align` attributes unconditionally.

We've done measurements with Miri and have determined that `noalias` shouldn't
break code. The requirements that allow us to add dereferenceable and align
have been long documented in the standard library documentation.

pcwalton committed:
Update documentation for `drop_in_place()`

pcwalton commented:
r? @oli-obk 

cc @RalfJung @saethlin

saethlin commented on library/core/src/ptr/mod.rs:
```diff
@@ -436,23 +436,41 @@ mod mut_ptr;
 ///
-/// * The value `to_drop` points to must be valid for dropping, which may mean it must uphold
-///   additional invariants - this is type-dependent.
+/// * While `drop_in_place` is executing, the only way to access parts of
```
@RalfJung Would it be better to say
```suggestion
/// * As soon as `drop_in_place` begins executing, the only way to access parts of
```

RalfJung replied:
What is the difference you are getting at here?

One `drop_in_place` returns, old references can be used again, at least as far as the aliasing model goes.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/3493072501/jobs/5847493939) [(plain)](https://github.com/rust-lang/rust/commit/02cfabe6710020df56c306b873cfa7f111504ac6/checks/9564174087/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
Attempting with retry: make prepare
---
---- src/ptr/mod.rs - ptr::drop_in_place (line 442) stdout ----
error[E0133]: dereference of raw pointer is unsafe and requires unsafe function or block
 --> src/ptr/mod.rs:445:21
  |
6 |     let mut value = &mut *to_drop;
  |                     ^^^^^^^^^^^^^ dereference of raw pointer
  |
  = note: raw pointers may be null, dangling or unaligned; they can violate aliasing rules and cause data races: all of these are undefined behavior
error: aborting due to previous error

For more information about this error, try `rustc --explain E0133`.
Couldn't compile the test.
Couldn't compile the test.

failures:
    src/ptr/mod.rs - ptr::drop_in_place (line 442)

test result: FAILED. 3916 passed; 1 failed; 36 ignored; 0 measured; 0 filtered out; finished in 48.08s

error: doctest failed, to rerun pass `-p core --doc`
```

</details>

pcwalton commented:
This patch reduces the number of stack-to-stack copies from 2.50% to 2.36%, a 6% decrease.

pcwalton committed:
Add missing "unsafe" to fix doctest

RalfJung commented:
The Miri version of this is https://github.com/rust-lang/rust/pull/103957. I'd prefer if we could land that first.

thomcc commented:
Just out of idle curiosity[^1]...

@bors try @rust-timer queue

[^1]: I don't expect tons but who knows how many of those stack/stack copies are in important places 👀

rust-timer commented:
Awaiting bors try build completion.

@rustbot label: +S-waiting-on-perf

<!-- rust-timer: temporary -->

bors commented:
:hourglass: Trying commit 53f21aaa99a33fbf8731818bb4826b62537f2986 with merge 7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a...
<!-- homu: {"type":"TryBuildStarted","head_sha":"53f21aaa99a33fbf8731818bb4826b62537f2986","merge_sha":"7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a"} -->

pcwalton commented:
> The Miri version of this is #103957. I'd prefer if we could land that first.

As long as it doesn't take too long. That PR has been sitting idle for 5 days. I don't want this to bitrot.

RalfJung commented:
Bitrot risk for this PR seems low and 5 days is not very long. I don't think this is an urgent change either.

RalfJung commented on library/core/src/ptr/mod.rs:
````diff
@@ -436,23 +436,41 @@ mod mut_ptr;
 /// # Safety
 ///
-/// Behavior is undefined if any of the following conditions are violated:
+/// Immediately upon executing, `drop_in_place` takes out a mutable borrow on the
+/// pointed-to-value. Effectively, this function is implemented like so:
+///
+/// ```
+/// # struct Foo { x: i32 }
+/// unsafe fn drop_in_place(to_drop: *mut Foo) {
+///     let mut value = &mut *to_drop;
+///     // ... drop the fields of `value` ...
+/// }
+/// ```
````
What we actually do is stronger than that: we have an `&mut` *function argument*. That makes a different because scope matters.

So if you want to write this in code, I'd suggest something like
````
/// ```
/// # struct Foo { x: i32 }
/// unsafe fn drop_in_place(to_drop: *mut Foo) {
///     drop_in_place_inner(&mut *to_drop);
///     unsafe fn drop_in_place_inner(oto_drop: &mut Foo) {
///         // ... drop the fields of `value` ...
///     }
/// }
/// ```
````

RalfJung commented on library/core/src/ptr/mod.rs:
```diff
@@ -436,23 +436,41 @@ mod mut_ptr;
 /// * `to_drop` must be [valid] for both reads and writes.
 ///
-/// * `to_drop` must be properly aligned.
+/// * `to_drop` must be properly aligned, even if T has size 0.
+///
+/// * `to_drop` must be nonnull, even if T has size 0.
```
```suggestion
/// * `to_drop` must be properly aligned, even if `T` has size 0.
///
/// * `to_drop` must be nonnull, even if `T` has size 0.
```

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/3495155516/jobs/5853547571)
Build commit: 7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a (`7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/3495155516/jobs/5853547571"},"merge_sha":"7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a"} -->

rust-timer commented:
Queued 7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a with parent 83356b78c4ff3e7d84e977aa6143793545967301, future [comparison URL](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a).
<!-- rust-timer: temporary -->

rust-timer commented:
Finished benchmarking commit (7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a): [comparison URL](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a&stat=instructions:u).

### Overall result: ❌✅ regressions and improvements - ACTION NEEDED


Benchmarking this pull request likely means that it is perf-sensitive, so we're automatically marking it as not fit for rolling up. While you can manually mark this PR as fit for rollup, we strongly recommend not doing so since this PR may lead to changes in compiler perf.

**Next Steps**: If you can justify the regressions found in this try perf run, please indicate this with `@rustbot label: +perf-regression-triaged` along with sufficient written justification. If you cannot justify the regressions please fix the regressions and do another perf run. If the next run shows neutral or positive results, the label will be automatically removed.

@bors rollup=never
@rustbot label: +S-waiting-on-review -S-waiting-on-perf +perf-regression


### [Instruction count](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a&stat=instructions:u)
This is a highly reliable metric that was used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.5%  | [0.4%, 0.7%]   | 7     |
| Regressions ❌ <br /> (secondary)  | 3.2%  | [0.5%, 9.5%]   | 7     |
| Improvements ✅ <br /> (primary)   | -0.6% | [-0.6%, -0.6%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | 0.4%  | [-0.6%, 0.7%]  | 8     |

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a&stat=max-rss)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean  | range          | count |
|:----------------------------------:|:-----:|:--------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -     | -              | 0     |
| Regressions ❌ <br /> (secondary)  | -     | -              | 0     |
| Improvements ✅ <br /> (primary)   | -0.9% | [-0.9%, -0.9%] | 1     |
| Improvements ✅ <br /> (secondary) | -     | -              | 0     |
| All ❌✅ (primary)                 | -0.9% | [-0.9%, -0.9%] | 1     |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=83356b78c4ff3e7d84e977aa6143793545967301&end=7250e7fb50ffb04be8243ed3f0d9baa1caa63d4a&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | 0.9% | [0.7%, 1.1%] | 7     |
| Regressions ❌ <br /> (secondary)  | 4.3% | [1.2%, 9.2%] | 5     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | 0.9% | [0.7%, 1.1%] | 7     |
</details>

RalfJung commented:
Some of that regression is probably caused by emitting more attributes.

pcwalton commented:
If folks are concerned about the performance regression, I can disable this in incremental compilation mode. It doesn't sit that well with me though—it's essentially just improving compilation performance by artificially making LLVM able to optimize less.

pcwalton commented:
Oh, also keep in mind that I'm testing against LLVM 16, while the benchmark presumably isn't.

thomcc commented:
> If folks are concerned about the performance regression, I can disable this in incremental compilation mode

I can't speak to anybody else, but I'm not concerned. Or, my concern is just that the value of this kind of thing can't really be measured using our existing perf suite, but that's an existing issue that shouldn't impact this PR.

oli-obk commented on compiler/rustc_ty_utils/src/abi.rs:
```diff
@@ -331,10 +344,16 @@ fn fn_abi_new_uncached<'tcx>(
     use SpecAbi::*;
     let rust_abi = matches!(sig.abi, RustIntrinsic | PlatformIntrinsic | Rust | RustCall);
 
+    let is_drop_in_place = match (cx.tcx.lang_items().drop_in_place_fn(), fn_def_id) {
+        (Some(drop_in_place_fn), Some(fn_def_id)) => drop_in_place_fn == fn_def_id,
+        _ => false,
+    };
```
```suggestion
    let is_drop_in_place = cx.tcx.lang_items().drop_in_place_fn() == Some(fn_def_id);
```

oli-obk commented:
@rustbot author

perf regression is entirely in LLVM, so since this causes actual runtime improvements, and the regressions are small in primary tests, this lgtm

RalfJung commented:
https://github.com/rust-lang/rust/pull/103957 finally landed, so from a Miri / MIR semantics perspective this is good to go now.

There are a bunch of review comments above though.

JohnCSimon commented:
> #103957 finally landed, so from a Miri / MIR semantics perspective this is good to go now.
> 
> There are a bunch of review comments above though.

@pcwalton  Can you please post your status on this PR? It has sat idle for months.

pnkfelix commented:
This ended up landing via PR #111807 if I understand correctly.
